### PR TITLE
Merge labels into trainer batches

### PR DESCRIPTION
**Pull request by acisseJZhong**
## Summary

Merge labels into the input dictionary at the `Trainer.batch_generator()`
boundary. Training paths now pass a single batch object through
`train_step()` and `forward_backward_step()`, while dataloaders keep returning
the existing `(input_dict, labels)` tuple and `preprocess_inputs()` keeps its
separate output contract.

The corresponding GraphTrainer, TorchFT, Flux, and test call sites are updated
to use the same batch contract.

## Testing

- Trainer, validator, invalid-loss, and chunked-loss tests: 33 passed, 2
  subtests passed.
- Targeted Pyrefly: 0 errors.
- Five deterministic training steps produced bitwise-identical loss and
  gradient norm before and after the refactor.

<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

### CODE REVIEW

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -741,46 +742,43 @@ def batch_generator(
         while True:
             data_load_start = time.perf_counter()
             try:
-                batch = next(data_iterator)
+                input_dict, labels = next(data_iterator)
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: `data_iterator` should output a `input_dict`, not a tuple

File: torchtitan/trainer.py [hunk 2]
```diff
@@ -741,46 +742,43 @@ def batch_generator(
...
-            yield input_dict, labels
+            yield {**input_dict, "labels": labels}
 
     @sl.log_trace_span("fwd_bwd")
     def forward_backward_step(
         self,
         *,
-        input_dict: dict[str, torch.Tensor] | list[dict[str, torch.Tensor]],
-        labels: torch.Tensor | list[torch.Tensor],
+        batch: dict[str, Any] | list[dict[str, Any]],
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: shouldn't call this batch, which is ambiguous -- can we still call it `input_dict`? in the sense that it's input to the trainer (not input tokens to the model)

btw when would it be a `list`?

Comment (acisseJZhong) [reply torchtitan/trainer.py]: It's a list when PP > 1.

File: torchtitan_recipes/tests/features.py [hunk 3]
```diff
@@ -44,13 +45,11 @@ def __init__(self, config: Config):
     def forward_backward_step(
         self,
         *,
-        input_dict: dict[str, torch.Tensor] | list[dict[str, torch.Tensor]],
-        labels: torch.Tensor | list[torch.Tensor],
+        batch: dict[str, Any] | list[dict[str, Any]],
```
Comment (tianyu-l) [inline torchtitan_recipes/tests/features.py]: nit: align this name as well, and pls check other files

File: torchtitan/trainer.py [hunk 4]
```diff
@@ -890,16 +884,16 @@ def train_step(self, data_iterator: Iterator[TrainerBatch]):
...
         # All groups form one optimizer step; each group feeds one fwd-bwd call.
-        microbatch_groups: list[list[TrainerBatch]] = []
+        microbatch_groups: list[list[dict[str, Any]]] = []
         local_valid_tokens = 0
         for _ in range(self.gradient_accumulation_steps):
             microbatches = []
             for _ in range(self.num_pp_microbatches):
                 with sl.log_trace_span("fetching_batch"):
-                    input_dict, labels = next(data_iterator)
+                    batch = next(data_iterator)
```
Comment (tianyu-l) [inline torchtitan/trainer.py]:
```suggestion
                    input_dict = next(data_iterator)
```
alternatively we could call it `microbatch`

**Review by tianyu-l — APPROVED** [review]
